streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -541,10 +555,82 @@ public void suspend() {
      *                               or if the task producer got fenced (EOS)
      */
     // visible for testing
-    void suspend(final boolean clean) {
-        closeTopology(); // should we call this only on clean suspend?
+    void suspend(final boolean clean,
+                 final boolean isZombie) {
+        try {
+            closeTopology(); // should we call this only on clean suspend?
+        } catch (final RuntimeException fatal) {
+            if (eosEnabled && !isZombie) {
+                if (transactionInFlight) {
+                    try {
+                        producer.abortTransaction();
```

Comment:
If suspend fails, we will always call `task.close(..)` again inside AssignedTasks, in which we will abort transactions and close record collectors. So do we have to do it here?

Reply:
That's a good point. However, we need to make sure that we don't fail a second time with `closeToplogy()` when `task.close(...)` calls `task.suspend(false,...)` a second time.

Will update the PR.